datafusion/physical-plan/src/sorts/sort.rs

```diff
@@ -1197,35 +1197,55 @@ impl ExecutionPlan for SortExec {
     ) -> Result<SendableRecordBatchStream> {
         trace!("Start SortExec::execute for partition {} of context session_id {} and task_id {:?}", partition, context.session_id(), context.task_id());
 
-        let mut input = self.input.execute(partition, Arc::clone(&context))?;
-
-        let execution_options = &context.session_config().options().execution;
-
-        trace!("End SortExec's input.execute for partition: {}", partition);
-
         let sort_satisfied = self
             .input
             .equivalence_properties()
             .ordering_satisfy_requirement(&LexRequirement::from(self.expr.clone()));
 
+        let input_exec = Arc::clone(&self.input);
+
+        let execution_options = &context.session_config().options().execution;
+
+        trace!("End SortExec's input.execute for partition: {}", partition);
+
         match (sort_satisfied, self.fetch.as_ref()) {
-            (true, Some(fetch)) => Ok(Box::pin(LimitStream::new(
-                input,
-                0,
-                Some(*fetch),
-                BaselineMetrics::new(&self.metrics_set, partition),
-            ))),
-            (true, None) => Ok(input),
+            (true, Some(fetch)) => {
+                let input = input_exec.execute(partition, Arc::clone(&context))?;
+                Ok(Box::pin(LimitStream::new(
+                    input,
+                    0,
+                    Some(*fetch),
+                    BaselineMetrics::new(&self.metrics_set, partition),
+                )))
+            }
+            (true, None) => self.input.execute(partition, Arc::clone(&context)),
             (false, Some(fetch)) => {
+                let schema = input_exec.schema();
                 let mut topk = TopK::try_new(
                     partition,
-                    input.schema(),
+                    schema,
                     self.expr.clone(),
                     *fetch,
                     context.session_config().batch_size(),
                     context.runtime_env(),
                     &self.metrics_set,
                 )?;
+                let input_exec = if context
+                    .session_config()
+                    .options()
+                    .optimizer
+                    .enable_dynamic_filter_pushdown
+                {
+                    // Try to push down the dynamic filter. If the execution plan doesn't
+                    // support it, push_down_filter will return None and we'll
+                    // keep the original input_exec.
+                    input_exec
```

Comment:
I still find it very strange that this pushdown happens *after* `execute` gets called -- would you be open to a PR that tries to do this with an optimizer pass? 

Apart from keeping the optimization and execution separate, which I think will help overall understandability, doing the pushdown as an optimizer pass will make the filters visible on explain plans (and thus easier to verify it was happening as expected)

Reply:
Okay, I'll see what I can do...

Reply:
I can try to help do this later in the week as well

Reply:
So I gave it a shot. It was a FIGHT and unfortunately I think I found the deal breaker way too late: `TopK` itself is only ever created in `SortExec::execute` and depends on the `partition` parameter to `execute`. So I don't see how we can get a reference to it's state during an optimizer pass. Even if we address this case I'm afraid that there will be other optimizations (joins?) that we then can't complete because we boxed ourselves into doing it as an optimizer pass instead of at runtime.